src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs

```diff
@@ -51,13 +53,20 @@ public static IMvcCoreBuilder AddMvcCore(this IServiceCollection services)
                 throw new ArgumentNullException(nameof(services));
             }
 
-            var partManager = GetApplicationPartManager(services);
+            var environment = GetServiceFromCollection<IWebHostEnvironment>(services);
+            var partManager = GetApplicationPartManager(services, environment);
             services.TryAddSingleton(partManager);
 
             ConfigureDefaultFeatureProviders(partManager);
             ConfigureDefaultServices(services);
             AddMvcCoreServices(services);
 
+            if (environment?.IsDevelopment() ?? false)
```

Comment:
Blazor WASM currently uses Debugging.IsSupported to determine this - https://github.com/dotnet/aspnetcore/blob/main/src/Components/Shared/src/HotReloadFeature.cs#L6-L16. But the experience is nicer there. The trimmer yanks out all of that code during publish, so it entirely no-ops. We don't have the benefit here. Also, aspnetcore's deployment is slightly centered around environments rather than build / publish.

@stephentoub, do you have suggestions? Would https://github.com/dotnet/runtime/issues/51159 solve this problem?

Reply:
> @stephentoub, do you have suggestions? Would dotnet/runtime#51159 solve this problem?

Not clear whether it would or not; it'd depend on how we implemented it.  Most of our IsSupported properties are initialized from AppContext (in addition to being recognized by the linker and relevant paths trimmed out entirely when the linker runs); I don't know if we have any that fall back to checking environment variables as well, though we do that for other settings elsewhere.  It'd also be possible to look for some other condition, but we'd need whatever it is to be very cheap, since we'd be paying it on pretty much every app startup.  @eerhardt, you've thought a bit about this; what would you recommend?

Reply:
We could predicate it on it being a Debug build I guess? I'm assuming there's metadata written to the app assembly in that case that we could check. I thought I saw another part of the Hot Reload toolchain doing something similar recently.

Reply:
The main benefit of an IsSupported property somewhere would be unifying that logic, whatever it checks, so that we don't have to replicate it in multiple places, and so that it can do the most efficient thing.  Unfortunately accessing metadata via reflection is rarely the most efficient thing :wink:

Reply:
All the more reason to not have to restart the process again :wink: \*cough Hot Restart\*

Reply:
@mikem8361, what is the runtime looking for to determine whether to load assemblies as hot reloadable?  Presumably a HotReload.IsSupported property (or whatever we chose to call it) could just surface the same answer the runtime's already computed?

Reply:
Using the existing `Debugger.IsSupported` switch, that we use elsewhere for trimming hot reload paths, doesn't seem like the best option here. If that switch isn't explicitly set, it defaults to `true` (since by default you can debug .NET app). That would mean the `HotReloadService` would get added all the time, even in Release published apps. It "works" for Blazor WASM because we explicitly set `Debugger.IsSupported` to `false` for non-Debug builds in the Blazor WASM SDK.

Does "Hot Reload" work in apps that didn't have a debugger attached when the process was started? For example, if I start my app, run for a few minutes and then attach a debugger, can I then start "hot reloading" things? If not, maybe an option here would be to check if a Debugger is attached, and only if it is add the HotReloadService?

But I agree with @stephentoub here that the most correct option feels like adding a dedicated HotReload.IsSupported property and surface the same value/logic that the runtime uses.

Reply:
> Does "Hot Reload" work in apps that didn't have a debugger attached when the process was started?

Hot reload works in apps that don't have the debugger attached, period.

Reply:
> Does "Hot Reload" work in apps that didn't have a debugger attached when the process was started? 

With dotnet-watch and the Ctrl-F5 scenarios in VS, a debugger isn't attached (and once a hot reload delta is applied, the runtime prevents attaching a debugger). 

The runtime currently requires setting `DOTNET_MODIFIABLE_ASSEMBLIES=Debug` for the ApplyDelta API to be called. I'm not a 100% sure if that env is also required with the debugger (probably not), but perhaps we could `HotReload.IsSupported => Debugger.IsAttached || Env["DOTNET_MODIFIABLE_ASSEMBLIES"] == "Debug"`?

Reply:
Yes, I think this would be the best way though I would say something `HotReload.IsSupported => Debugger.IsAttached || Env["DOTNET_MODIFIABLE_ASSEMBLIES"] != "" || Env["COMPlus_ForceEnc"] == "1"`. There is one problem with this logic: when the env vars are NOT set and a debugger attaches (a typical scenario) this proposed property would incorrectly return true.

ENC for a module is enabled on launch by the debugger on module load callbacks.  These env vars are only needed if you want ENC when the debugger attaches. 

If we want this global property part of the runtime, it probably should be renamed to something without "HotReload". We have been avoiding that term in the public APIs (see AssemblyExtensions.ApplyUpdate or AssemblyExtensions.GetApplyUpdateCapabilities).  

The runtime itself couldn't do any better with some kind of QCall because it doesn't have a concept of a global "is ENC'able or hotreloadable" flag. It uses these env vars (or a call to a debugging API) to create special ENC per-module data structures. The runtime could return that an assembly/module is ENC'able/hotreloadable but something global is needed.

Reply:
Just a minor difference in behavior with Mono: right now we always require `Env["DOTNET_MODIFIABLE_ASSEMBLIES"]!=""` to be set at startup (possibly it's enough for the variable to be set just before any assembly we intend to modify executes any code, but I haven't tested this).  The mono debugger engine doesn't have any mechanism to say "enable EnC for a module" at the moment.

Reply:
But actually checking the environment variable is a necessary but not sufficient condition for whether we _actually_ have hot reload at runtime - with the ongoing [componentization of Mono](https://github.com/dotnet/runtime/issues/50862), workloads can decide whether to bundle hot reload support in some configurations or builds of the user's app.

So framework or library authors really need a separate feature switch if they want to affect runtime behavior. (I think we're all aware of the open issue for that, but just in case it's https://github.com/dotnet/runtime/issues/51159)

Reply:
I filed https://github.com/dotnet/aspnetcore/issues/33477 to track updating this code once we've had a chance to figure out what the right semantics should be for this flag. Given solving this is largely a hygiene issue, I'm not going to block on resolving if we're able to get support for custom attributes working end-to-end in preview6.